doc/dev/custom_devices.rst

```diff
@@ -0,0 +1,159 @@
+            {
+            }
+
+            virtual void MatrixOperation(const std::vector<std::complex<double>> &,
+                                         const std::vector<QubitIdType> &, bool)
+            {
+            }
+
+            virtual auto Expval(ObsIdType) -> double { return 0.0; }
+            virtual auto Var(ObsIdType) -> double { return 0.0; }
+            virtual void State(DataView<std::complex<double>, 1> &) {}
+            virtual void Probs(DataView<double, 1> &) {}
+            virtual void PartialProbs(DataView<double, 1> &, const std::vector<QubitIdType> &) {}
+            virtual void Sample(DataView<double, 2> &, size_t) {}
+            virtual void PartialSample(DataView<double, 2> &, const std::vector<QubitIdType> &, size_t) {}
+            virtual void Counts(DataView<double, 1> &, DataView<int64_t, 1> &, size_t) {}
+
+            virtual void PartialCounts(DataView<double, 1> &, DataView<int64_t, 1> &,
+                                       const std::vector<QubitIdType> &, size_t)
+            {
+            }
+
+            virtual void Gradient(std::vector<DataView<double, 1>> &, const std::vector<size_t> &) {}
+        };
+
+In addition to implementing the ``QuantumDevice`` class, one must implement the following method:
+
+.. code-block:: c++
+
+    extern "C" Catalyst::Runtime::QuantumDevice*
+    getCustomDevice() { return new CustomDevice(); }
+
+where ``CustomDevice()`` is a constructor for your custom device.
+``CustomDevice``'s destructor will be called by the runtime.
+
+.. note::
+
+    This interface might change quickly in the near future.
+    Please check back regularly for updates and to ensure your device is compatible with a specific version of Catalyst.
+
+How to compile custom devices
+=============================
+
+One can follow the ``catalyst/runtime/tests/third_party/CMakeLists.txt`` `as an example. <https://github.com/PennyLaneAI/catalyst/blob/26b412b298f22565fea529d2019554e7ad9b9624/runtime/tests/third_party/CMakeLists.txt>`_
+
+.. code-block:: cmake
+
+        cmake_minimum_required(VERSION 3.20)
+
+        project(third_party_device)
+
+        set(CMAKE_CXX_STANDARD 20)
+        set(CMAKE_CXX_STANDARD_REQUIRED ON)
+
+        add_library(dummy_device SHARED dummy_device.cpp)
+        target_include_directories(dummy_device PUBLIC ${runtime_includes})
+        set_property(TARGET dummy_device PROPERTY POSITION_INDEPENDENT_CODE ON)
+
+
+
+Integration with Python devices
+===============================
+
+If you already have a custom PennyLane device defined in Python and have added a shared object that corresponds to your implementation of the ``QuantumDevice`` class, then all you need to do is to add a ``get_c_interface`` method to your PennyLane device.
+The ``get_c_interface`` method should be a static method that takes no parameters and returns the complete path to your shared library with the ``QuantumDevice`` implementation.
```

Comment:
Defining this static method makes me worry about the package portability! What do you think of defining a default directory path somewhere in the frontend or runtime for available shared libraries? For example, it can be setup in a way that if this `get_c_interface` fails then it'll lookup the library from the default devices directory (this can be useful for our current drivers).

Reply:
Can you elaborate why is this static method a worry about package portability?

Reply:
Technically, in UNIX at least, these libraries could be installed in any of the `LD_LIBRARY_PATH` or similar and it would still succeed. See dlopen:

```
       If filename is NULL, then the returned handle is for the main
       program.  If filename contains a slash ("/"), then it is
       interpreted as a (relative or absolute) pathname.  Otherwise, the
       dynamic linker searches for the object as follows (see [ld.so(8)](https://man7.org/linux/man-pages/man8/ld.so.8.html)
       for further details):

       •  (ELF only) If the calling object (i.e., the shared library or
          executable from which dlopen() is called) contains a DT_RPATH
          tag, and does not contain a DT_RUNPATH tag, then the
          directories listed in the DT_RPATH tag are searched.

       •  If, at the time that the program was started, the environment
          variable LD_LIBRARY_PATH was defined to contain a colon-
          separated list of directories, then these are searched.  (As a
          security measure, this variable is ignored for set-user-ID and
          set-group-ID programs.)

       •  (ELF only) If the calling object contains a DT_RUNPATH tag,
          then the directories listed in that tag are searched.

       •  The cache file /etc/ld.so.cache (maintained by [ldconfig(8)](https://man7.org/linux/man-pages/man8/ldconfig.8.html)) is
          checked to see whether it contains an entry for filename.

       •  The directories /lib and /usr/lib are searched (in that
          order).
```

I suppose it is possible to remove / update this early check to mimic the same dlopen here.

Reply:
My concern is with our current devices (lightning sims and OQ3). As discussed offline, we can address it in a follow up PR. :+1: